handler/src/main/java/io/netty/handler/ssl/ReleasableOpenSslEngine.java

```diff
@@ -0,0 +1,1988 @@
+        // There won't be any application data until we're done handshaking.
+        // We first check handshakeFinished to eliminate the overhead of extra JNI call if possible.
+        return handshakeState == HandshakeState.FINISHED ? SSL.pendingReadableBytesInSSL(ssl) : 0;
+    }
+
+    private SSLEngineResult newResult(
+            int bytesConsumed, int bytesProduced, SSLEngineResult.HandshakeStatus status) throws SSLException {
+        return new SSLEngineResult(
+                getEngineStatus(), mayFinishHandshake(status != FINISHED ? getHandshakeStatus() : status)
+                , bytesConsumed, bytesProduced);
+    }
+
+    private void closeAll() throws SSLException {
+        receivedShutdown = true;
+        closeOutbound();
+        closeInbound();
+    }
+
+    private void rejectRemoteInitiatedRenegation() throws SSLHandshakeException {
+        if (rejectRemoteInitiatedRenegation && SSL.getHandshakeCount(ssl) > 1) {
+            // TODO: In future versions me may also want to send a fatal_alert to the client and so notify it
+            // that the renegotiation failed.
+            shutdown();
+            throw new SSLHandshakeException("remote-initiated renegotation not allowed");
+        }
+    }
+
+    public final SSLEngineResult unwrap(final ByteBuffer[] srcs, final ByteBuffer[] dsts) throws SSLException {
+        return unwrap(srcs, 0, srcs.length, dsts, 0, dsts.length);
+    }
+
+    private ByteBuffer[] singleSrcBuffer(ByteBuffer src) {
+        singleSrcBuffer[0] = src;
+        return singleSrcBuffer;
+    }
+
+    private void resetSingleSrcBuffer() {
+        singleSrcBuffer[0] = null;
+    }
+
+    private ByteBuffer[] singleDstBuffer(ByteBuffer src) {
+        singleDstBuffer[0] = src;
+        return singleDstBuffer;
+    }
+
+    private void resetSingleDstBuffer() {
+        singleDstBuffer[0] = null;
+    }
+
+    @Override
+    public final synchronized SSLEngineResult unwrap(
+            final ByteBuffer src, final ByteBuffer[] dsts, final int offset, final int length) throws SSLException {
+        try {
+            return unwrap(singleSrcBuffer(src), 0, 1, dsts, offset, length);
+        } finally {
+            resetSingleSrcBuffer();
+        }
+    }
+
+    @Override
+    public final synchronized SSLEngineResult wrap(ByteBuffer src, ByteBuffer dst) throws SSLException {
+        try {
+            return wrap(singleSrcBuffer(src), dst);
+        } finally {
+            resetSingleSrcBuffer();
+        }
+    }
+
+    @Override
+    public final synchronized SSLEngineResult unwrap(ByteBuffer src, ByteBuffer dst) throws SSLException {
+        try {
+            return unwrap(singleSrcBuffer(src), singleDstBuffer(dst));
+        } finally {
+            resetSingleSrcBuffer();
+            resetSingleDstBuffer();
+        }
+    }
+
+    @Override
+    public final synchronized SSLEngineResult unwrap(ByteBuffer src, ByteBuffer[] dsts) throws SSLException {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Consider using varargs for methods or constructors which take an array the last parameter. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)